cpp/src/arrow/type.h

```diff
@@ -26,9 +26,12 @@
 #include <utility>
 #include <vector>
 
+#include "arrow/result.h"
 #include "arrow/type_fwd.h"  // IWYU pragma: export
 #include "arrow/util/checked_cast.h"
 #include "arrow/util/macros.h"
+#include "arrow/util/small_vector.h"
+#include "arrow/util/variant.h"
```

Comment:
Hmm, it's a pity to start pulling the `variant` header in such a widely-used header...

Reply:
https://issues.apache.org/jira/browse/ARROW-8113